doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,1253 @@
+permission of the user to spawn an elevated client application. The user would
+see a UAC prompt, they'd accept, and then they'd be able to have an elevated
+shell alongside their unelevated tabs.
+
+However, this creates an escalation of privilege vector. Now, there's an
+unelevated window which is connected directly to an elevated process. At this
+point, any other unelevated application could send input to the Terminal's
+`HWND`, making it possible for another unelevated process to "drive" the
+Terminal window and send commands to the elevated client application.
+
+### Drag and drop tabs to create new windows
+
+Another important scenario we're targeting is the ability to drag a tab out of
+the Terminal window and create a new Terminal window. Obviously, when we do
+this, we want the newly created window to be able to persist all the same state
+that the original window had for that tab. For us, that primarily means the
+buffer contents and connection state.
+
+However, _how_ do we move the terminal state to another window? The terminal
+state is all located in-memory of the thread that created the `TermControl`
+hosting the terminal buffer. It's fairly challenging to re-create this state in
+another process.
+
+Take a look at [this
+document](https://github.com/windows-toolkit/Sample-TabView-TearOff) for more
+background on how the scenario might work.
+
+There's really only a limited selection of things that a process could transfer
+to another with a drag and drop operation. If we wanted to use a string to
+transfer the data, we'd somehow need to serialize then entire state of the tab,
+it's tree of panes, and the state of each of the buffers in the tab, into some
+sort of string, and then have the new window deserialize that string to
+re-create the tab state. Consider that each buffer might include 32000 rows of
+80+ characters each, each with possibly RGB attributes, and you're looking at
+30MB+ of raw data to serialize and de-serialize per buffer, minimum. This sounds
+extremely fragile, if not impossible to do robustly.
+
+What we need is a more effective way for separate Terminal windows to to be able
+to connect to and display content that's being hosted in another process.
+
+## Solution Design
+
+### Window and Content Processes
+
+To begin, let's first take a look at a rough diagram of how the Windows Terminal
+process looks today:
+
+![figure-001](figure-001.png)
+
+Currently, the entire Windows Terminal exists as a single process composed of
+many parts. it has a top Win32 layer responsible for the window, which includes
+a UWP XAML-like App layer, which embeds many `TermControl`s, each of which
+contains the buffer and renderer, and communicates with a connection to another
+process.
+
+The primary concept introduced by this spec is the idea of two types of process,
+which will work together to create a single Terminal window. These processes
+will be referred to as the "Window Process" and the "Content Process".

```

Comment:
This resembles the architecture of modern text editors like Neovim and Xi-Editor, which separates the editor backend and the frontend.

Reply:
Come to think about it, the terminal actually resembles text editors in many ways like custom text rendering, content refreshing, input & output. I wonder if WT will be the first terminal that utilize the frontend/backend methodology.

Reply:
Alas! Rxvt-unicode has a client/server architecture that you can optionally use as well.